transport-native-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollStreamChannel.java

```diff
@@ -876,6 +876,25 @@ void epollInReady() {
                     allocHandle.incMessagesRead(1);
                     pipeline.fireChannelRead(byteBuf);
                     byteBuf = null;
+
+                    if (fd().isInputShutdown()) {
+                        // We need to do this for two reasons:
+                        //
+                        // - If the input was shutdown in between (which may be the case when the user did it in the
+                        //   fireChannelRead(...) method we should not try to read again to not produce any
+                        //   miss-leading exceptions.
+                        //
+                        // - If the user closes the channel we need to ensure we not try to read from it again as
+                        //   the filedescriptor may be re-used already by the OS if the system is handling a lot of
+                        //   concurrent connections and so needs a lot of filedescriptors. If not do this we risk
+                        //   reading data from a filedescriptor that belongs to another socket then the socket that
+                        //   was "wrapped" by this Channel implementation.
+                        //
+                        // We only need to do this in the epoll transport (and not the NIO) as the NIO
+                        // SocketChannelImpl.read(...) method will automatically return 0 when it was closed
```

Comment:
it will return `-1` (EOF) i think?

Also I can't think of a reason why we shouldn't have this for NIO. Otherwise we will force a close of the channel even if it is just the "input" side of the channel that has been closed.